instrumentation/okhttp/okhttp-3.0/library/README.md

```diff
@@ -0,0 +1,67 @@
+# Manual Instrumentation for OkHttp3
+
+Provides OpenTelemetry instrumentation for [okhttp3](https://square.github.io/okhttp/).
+
+## Quickstart
+
+### Add these dependencies to your project:
+
+Replace `OPENTELEMETRY_VERSION` with the latest stable
+[release](https://mvnrepository.com/artifact/io.opentelemetry). `Minimum version: 1.4.0`
+
+For Maven, add to your `pom.xml` dependencies:
+
+```xml
+
+<dependencies>
+  <dependency>
+    <groupId>io.opentelemetry.instrumentation</groupId>
+    <artifactId>opentelemetry-okhttp-3.0</artifactId>
+    <version>OPENTELEMETRY_VERSION</version>
+  </dependency>
+</dependencies>
+```
+
+For Gradle, add to your dependencies:
+
+```groovy
+implementation("io.opentelemetry.instrumentation:opentelemetry-okhttp-3.0:OPENTELEMETRY_VERSION")
+```
+
+### Usage
+
+The instrumentation library provides an OkHttp `Interceptor` implementation which instruments http
+client calls, including context propagation on the outgoing request, and span attributes based on
+the OpenTelemetry semantic conventions.
+
+In addition, it is straightforward to add support for in-process context propagation, based on standard
+OpenTelemetry APIs.
+
+```java
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.instrumentation.okhttp.v3_0.OkHttpTracing;
+import okhttp3.Dispatcher;
+import okhttp3.Interceptor;
+import okhttp3.OkHttpClient;
+
+import java.util.concurrent.ExecutorService;
+
+public class OkHttpConfiguration {
+
+  public Interceptor createInterceptor(OpenTelemetry openTelemetry) {
+    return OkHttpTracing.newBuilder(openTelemetry).build().newInterceptor();
+  }
+
+  public Dispatcher createDispatcher(ExecutorService executorService) {
+    return new Dispatcher(Context.taskWrapping(executorService));
```

Comment:
Unfortunately this doesn't seem to be enough to always get the correct context into callbacks. `OkHttp3Test` library tests use a similar construct, but `high concurrency test with callback` test still fails. In `okhttp:3.0.0` dispatcher will start queuing calls when too many of them are running at the same time so if you only make a few request this might work.

Reply:
Hmm. How is it on the 4.x line?

Does this mean that there is no way to get propagation to work right into okhttp async calls without bytecode munging?

Reply:
4.9.1 fails the same way, FYI.

I still think this is probably the best we can do for now. Should we just include a very large caveat about this issue, or would it better to not even recommend that people try to propagate the context into async calls?

Reply:
I think it's probably better not to recommend that - as a user, you usually don't know if your application never crosses that magical 64 concurrent requests number, and if it happens to do that you'll get unreliable traces.

Reply:
In the SIG meeting yesterday evening/morning, we talked through this, and Brave has a better approach to this library instrumentation that I'll be starting to figure out how to port over to otel, which shouldn't have the same mis-parenting issues that the Dispatcher approach ends up with.

Reply:
And, I'm going to update this PR to warn about this approach, rather than recommending it. Although, with our current instrumentation, there is no way to reliably get context propagation to work. You either get none, period, or you get potentially broken/mis-parented traces/spans.

Reply:
I also updated this PR to include some new instrumentation which does work with the `Call.Factory` approach that was used in Brave.